### PR TITLE
Add a ready mechanism to the conf pkg

### PR DESCRIPTION
**Pull request by jhchabran**
Fixes https://github.com/sourcegraph/sourcegraph/issues/29222

Originally, the issue mentions fixing the call sites to avoid having `conf` function calls in `init` packages, but this is a tentacular set of changes and introduce complexity, as all exported function from a package relying on `conf` package must be able to register the `Watch` hook to react on configuration changes.

So with @vrto, we took a different route: it's semantically perfectly correct to register a `conf.Watch` in an `init` function of a given package. 

What is incorrect (and causes trouble) is that the `conf.Watch` function will immediately fire an HTTP request to poll the config, regardless of where the application is in its initialization process. This is what led to the problem described in the original issue. 

Our proposed solution is to instead delay the configuration polling until the service is starting, ie the `main` function. To achieve this, we have updated the `conf.Init` function to be the one starting the continuously polling goroutine, if and only if we are in client mode. 

The resulting changes are very minimal and solve the original problem of unexpectedly firing HTTP requests by simply importing a package, (`sg` being a recurring place where this happened). 

It also makes the configuration pulling more logical to comprehend until we get the chance of refactoring this package entirely as it's rather confusing in its present form. 

The tests themselves are not impacted in any way. We still need the empty mechanism for those, so the test workaround has been left in place. 

## Test plan

- CI runs
- Manual QA in local:
	1. `sg start oss` 
	2. Added a print statement in a `conf.Watch` in `gitserver`
	3. Observed that the app is reloaded by `sg` because it changed 
	4. Observed the print statement output in the logs 
	5. Opened a browser, updated the site config 
	6. Observed the print statement being displayed again
		- This shows that the config package is still able to pull changes as intended even with the delayed start. 
- Check for every application and make sure that `conf.Init` is being called during application start.

### CODE REVIEW

File: internal/conf/init.go [hunk 1]
```diff
@@ -7,8 +7,15 @@ import (
 )
 
 func Init() {
+	// The default client is started in InitConfigurationServerFrontendOnly in
+	// the case of server mode.
+	if getMode() == modeClient {
```
Comment (bobheadxi) [inline internal/conf/init.go:12]: Does this server mode mean "server of site configuration"?

Comment (bobheadxi) [reply internal/conf/init.go:12]: yes! I also just read the docstring on `modeClient` which was very explanatory, should have gone to that first 🤣 Thanks!

File: internal/conf/init.go [hunk 2]
```diff
@@ -7,8 +7,15 @@ import (
 )
 
 func Init() {
```
[the remark this answers is not in the document]
Comment (jhchabran) [reply internal/conf/init.go:9]: We've made sure every service is calling those, but we should indeed add some docstring there! @vrto wanna take it?

[the remark this answers is not in the document]
Comment (vrto) [reply internal/conf/init.go:11]: @jhchabran took a stab :)

File: internal/conf/init.go [hunk 3]
```diff
@@ -6,9 +6,18 @@ import (
 	"github.com/sourcegraph/sourcegraph/internal/httpcli"
 )
 
+// Init function must be called by every service that requires access to components in conf package.
+// If the caller is in the client mode, we start a goroutine continuously polling for config changes.
```
Comment (jhchabran) [inline internal/conf/init.go:10]:
```suggestion
// Init function completes the initialization process of the conf package, starting the configuration continuous changes polling if in client mode. The conf.Watch function can safely be called before calling Init to register callbacks reacting to the changes. 
// 
// The Init function must be called early in an application initialization process, but tests do not need to call it. 
```

Reasoning about the changes 😊: 

- The fact it uses a goroutine is an implementation detail, so we don't need to talk about it. 
- It's quite common to start a docstring in Go with what the function does, then to state if there are any contextual things to know, which is why I moved the fact that it needs to be called by all apps at the end. 
- It not be clear for the reader what can be called before this, which is why we explicit the `conf.Watch` part. Same logic with the tests!

Comment (vrto) [reply internal/conf/init.go:10]: Love it, thanks!

**Comment by jhchabran** [comment]
@tsenart do you want to take a look before we merge this?

**Review by tsenart — APPROVED** [review]

**Comment by chrismwendt** [comment]
This seems to have broken the symbols service on dogfood, reverting in https://github.com/sourcegraph/sourcegraph/pull/32420

**Pull request merged**